src/math/spherical_harmonics.cc

```diff
@@ -181,55 +176,215 @@ namespace rascal {
      *
      * @param max_angular Compute up to this angular momentum number (l_max)
      *
-     * @return  (Eigen)vector containing the results.
-     *          Sized (l_max+1)**2, contains the l,m components in compressed
-     *          format, i.e. (00)(1-1)(10)(11)(2-2)....
+     * @return  (Eigen)matrix containing the results.
+     *          Sized (l_max+1)^2, the index collects l and m quantum numbers
+     *          stored in compact format (m varies fastest, from -l to l,
+     *          and l from 0 to l_max).
      */
     Vector_t compute_spherical_harmonics(
         const Eigen::Ref<const Eigen::Vector3d> & direction,
         size_t max_angular) {
-      using math::pow;
+      using std::pow;
       using std::sqrt;
-
+      Eigen::Vector3d my_direction;
+      if (std::abs((direction[0] * direction[0] + direction[1] * direction[1] +
+                    direction[2] * direction[2]) -
+                   1.0) > math::dbl_ftol) {
+        std::cerr << "Warning: math::compute_spherical_harmonics(): ";
+        std::cerr << "Direction vector unnormalized, normalizing it now";
+        std::cerr << std::endl;
+        my_direction = direction / direction.norm();
+      } else {
+        my_direction = direction;
+      }
       // The cosine against the z-axis is just the z-component of the
       // direction vector
-      double cos_theta = direction[2];
-      // The less efficient, but more intuitive implementation:
-      // double phi = std::atan2(direction[1], direction[0]);
-      double sqrt_xy = std::hypot(direction[0], direction[1]);
+      double cos_theta = my_direction[2];
+      double sqrt_xy = std::hypot(my_direction[0], my_direction[1]);
       // For a vector along the z-axis, define phi=0
       double cos_phi{1.0}, sin_phi{0.0};
       if (sqrt_xy >= math::dbl_ftol) {
-        cos_phi = direction[0] / sqrt_xy;
-        sin_phi = direction[1] / sqrt_xy;
+        cos_phi = my_direction[0] / sqrt_xy;
+        sin_phi = my_direction[1] / sqrt_xy;
       }
-      Vector_t harmonics =
-          Vector_t::Zero((max_angular + 1) * (max_angular + 1));
-      Matrix_t assoc_legendre_polynom =
+
+      Eigen::VectorXd harmonics =
+          Eigen::VectorXd::Zero(pow(max_angular + 1, 2));
+      Eigen::MatrixXd assoc_legendre_polynom =
           compute_assoc_legendre_polynom(cos_theta, max_angular);
-      MatrixX2_t cos_sin_m_phi =
+      Eigen::MatrixXd cos_sin_m_phi =
           compute_cos_sin_angle_multiples(cos_phi, sin_phi, max_angular);
 
-      size_t lm_base{0};  // starting point for storage
+      size_t l_block_index{0};
```

Comment:
For each value of `l`, we store a section of `2l+1` numbers. This variable is an index that tells us where, in the compact array, we store that section. Notice it increases by `2l+1` each iteration (over `l`).

Reply:
This construct appears in many sections of the code, as we need often to loop over l and have (rightly) switched from storing a lmax+1 * 2lmax+1 array with zeros to a (lmax+1)^2 full array. Was wondering if we should implement some kind of special iterator machinery to achieve this. at least, we should uniform the naming scheme. IMO for instance max_angular is not clearer than l_max, and angular_l could be just l, and so on and so forth. Seriously, whoever will look at this code should be familiar with spherical harmonics and the notation there is very well established.

Reply:
No single-letter variable names please, and 'l' is just terrible. We can use something like `l_max` if you like, but I'd only change it to differentiate from `max_radial` (which should really be `num_radial_channels`, because there are in fact `max_radial` radial components -- as opposed to the angular channel, where there are `max_angular + 1` components).

Reply:
IDK. In my opinion, in all of these math-heavy files, conforming with the mathematical notation that should eventually go in the curated section of the manual outweights general concerns such as "l is indistinguishable from I in sans fonts". Personally, if I have to see side-by-side a series of expressions containing l_max, l, m and the code, I find the code more readable if it uses the same names. Much as with the Mxpy notation Felix uses for the confluent hypergeometric function - that is the notation used in the handbook of mathematical functions and it makes sense to keep it, perhaps with a note saying where the notation comes from.

Reply:
Ugh, okay, fine. Since it seems to be the convention in a quasi-random sampling of spherical harmonics _codes_, we can change it to `l`/`l_max` and `m` - but in the cosmetics branch, not here. And `n` (index for the radial basis) needs to retain the `radial` moniker because it's otherwise frustratingly ambiguous; I'm even having second thoughts about the `m` appearing in the same context because it could easily be mistaken to have something to do with `n`.

Reply:
no absolutely strong feelings eh - but at the very least we need uniform naming conventions